spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+{
+	if (lut_seq >= NUM_LUT_SEQ) {
+		return -1;
+	}
+	if (*QSPI_LCKCR_ADDR & QSPI_LCKCR_LOCK) {
+		*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+		*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_UNLOCK;
+	}
+
+	for (int i = 0; i < 4; i++) {
+		*QSPI_LUTn_ADDR(QSPI_SEQ_START_REGN(lut_seq) + i) = (((uint32_t)lut->instrs[2 * i + 1]) << 16) | lut->instrs[2 * i];
+	}
+
+	*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+	*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_LOCK;
+}
+
+static void set_IPCR(int seq_num, size_t idatsz)
+{
+	*QSPI_IPCR_ADDR = QSPI_IPCR_SET_IDATSZ(QSPI_IPCR_SET_SEQID(*QSPI_IPCR_ADDR, seq_num), idatsz & 0xFFFF);
+}
+
+void qspi_setTCSH(uint8_t cycles)
+{
+	*QSPI_FLSHCR_ADDR = (*QSPI_FLSHCR_ADDR & ~(0x0F << 8)) | (cycles << 8);
+}
+
+void qspi_setTCSS(uint8_t cycles)
+{
+	*QSPI_FLSHCR_ADDR = (*QSPI_FLSHCR_ADDR & ~(0x0F)) | cycles;
+}
+
+int _qspi_readBusy(qspi_dev_t dev, unsigned int lut_seq, uint32_t addr, void *buf, size_t size)
+{
+	uint8_t to_read, byte, i;
+	uint16_t len = 0;
+	uint32_t reg;
+	int err;
+
+
+	if (dev != qspi_flash_a && dev != qspi_flash_b) {
+		return -ENODEV;
+	}
+	if (lut_seq >= NUM_LUT_SEQ || size > MAX_READ_LEN) {
+		return -1;
+	}
+
+	*QSPI_RBCT_ADDR = (*QSPI_RBCT_ADDR & ~(0x1F) | WATERMARK);
+
+	*QSPI_SFAR_ADDR = qspi_common.flash_base_addr[dev] + addr;
+
+	*QSPI_MCR_ADDR |= QSPI_MCR_CLR_RXF;
+
+	set_IPCR(lut_seq, size);
+
+	while (len < size) {
+		do {
+			reg = *QSPI_SR_ADDR;
+		} while ((reg & QSPI_SR_RXWE) == 0 && reg & QSPI_SR_BUSY);
+
+		for (i = 0; i < WATERMARK + 1 && len < size; i++) {
+			reg = *QSPI_RBDRn_ADRR(i);
+			for (byte = 0; byte < 4 && len + byte < size; byte++) {
+				((uint8_t *)buf)[len + byte] = reg & 0xFF;
+				reg >>= 8;
+			}
+			len += byte;
+		}
+		*QSPI_FR_ADDR |= QSPI_FR_RBDF; /* Buffer pop. */
+	}
+
+	while (*QSPI_SR_ADDR & QSPI_SR_BUSY)
+		;
+	*QSPI_MCR_ADDR |= QSPI_MCR_CLR_RXF;
+
+	return EOK;
+}
+
+static size_t write_tx(const void *data, size_t size)
+{
+	size_t byte, sent = 0;
+	uint32_t reg;
+
+	while ((*QSPI_SR_ADDR & QSPI_SR_TXFULL) == 0 && sent < size) {
+		reg = 0;
+		for (byte = 0; byte < 4 && sent + byte < size; byte++) {
+			reg |= ((uint8_t *)data)[sent + byte] << (8 * byte);
+		}
+		*QSPI_TBDR_ADDR = reg;
+		sent += byte;
+	}
+
+	return sent;
+}
+
+int _qspi_write(qspi_dev_t dev, unsigned int lut_seq, uint32_t addr, const void *data, size_t size)
+{
+	size_t sent = 0;
+	int err;
+	if (dev != qspi_flash_a && dev != qspi_flash_b) {
+		return -ENODEV;
+	}
+	if (lut_seq >= NUM_LUT_SEQ || size > MAX_READ_LEN) {
+		return -1;
+	}
+
+	if (*QSPI_SR_ADDR & QSPI_SR_TXEDA) /* Check if TX buffer is not empty. */
+		*QSPI_MCR_ADDR |= QSPI_MCR_CLR_TXF;
+
+	*QSPI_SFAR_ADDR = qspi_common.flash_base_addr[dev] + addr;
+
+	sent += write_tx(data, size);
+
+	set_IPCR(lut_seq, size);
+
+	while (sent < size) {
+		sent += write_tx(data + sent, size - sent);
+	}
```

Comment:
`write_tx` does this while already
```suggestion
	sent += write_tx(data + sent, size - sent);
```

Reply:
it doesn't as it checks if TX is full

Reply:
then I guess
```suggestion
	if (sent < size) {
		sent += write_tx(data + sent, size - sent);
	}
```

Reply:
then if our write requires more than 2 calls write_tx it won''t succeed, maybe I should change write_tx function somehow?